### PR TITLE
Support configuring SSL separately for remote access port

### PR DESCRIPTION
**Pull request by ywangd**
Though the remote access is implemented with the transport profile. Its configuration and behaviour should not be tied to the default transport profile like other profiles do. Users should be able to enable or disable SSL separately for the remote access port and configure differnt values for all SSL settings. These settings can also have different defaults.

This PR implements the above by:
* Adds a new xpack.security.remote_cluster.ssl.enabled setting to control whether SSL is enabled separately for the remote access port
* The above enabled setting defaults to true (unlike the default for tranport SSL)
* Client auth defaults to none for the remote access port
* Separate server SSL configuration validation

The PR also moves the remote access profile to be built first for more consistent error message.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/transport/TcpTransport.java [hunk 1]
```diff
@@ -1018,6 +1018,11 @@ public final TransportStats getStats() {
      */
     public static Set<ProfileSettings> getProfileSettings(Settings settings) {
         HashSet<ProfileSettings> profiles = new HashSet<>();
+        // Process remote cluster port first so that errors are consistently reported if there
+        // is direct usage of the _remote_cluster profile
+        if (REMOTE_CLUSTER_PORT_ENABLED.get(settings)) {
+            profiles.add(RemoteClusterPortSettings.buildRemoteAccessProfileSettings(settings));
+        }
```
Comment (ywangd) [inline server/src/main/java/org/elasticsearch/transport/TcpTransport.java:1025]: This is a minor improvement so that we don't get error messages like 
> profile [_remote_cluster] has no port configured

which is confusing because it seems that direct usage of the `_remote_cluster` profile is allowed.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/SSLService.java [hunk 2]
```diff
@@ -881,6 +902,15 @@ private static Settings getHttpTransportSSLSettings(Settings settings) {
         return builder.build();
     }
 
+    private static Settings getRemoteClusterSslSettings(Settings settings) {
+        final Settings remoteClusterSslSettings = settings.getByPrefix(XPackSettings.REMOTE_CLUSTER_SSL_PREFIX);
+        final Settings.Builder builder = Settings.builder().put(remoteClusterSslSettings);
+        if (builder.get("client_authentication") == null) {
+            builder.put("client_authentication", XPackSettings.REMOTE_CLUSTER_CLIENT_AUTH_DEFAULT);
+        }
```
Comment (ywangd) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/SSLService.java:910]: SSL for remote cluster port defaults to no client auth.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/SSLService.java [hunk 3]
```diff
@@ -591,8 +592,8 @@ static Map<String, Settings> getSSLSettingsMap(Settings settings) {
         sslSettingsMap.put(WatcherField.EMAIL_NOTIFICATION_SSL_PREFIX, settings.getByPrefix(WatcherField.EMAIL_NOTIFICATION_SSL_PREFIX));
         sslSettingsMap.put(XPackSettings.TRANSPORT_SSL_PREFIX, settings.getByPrefix(XPackSettings.TRANSPORT_SSL_PREFIX));
         sslSettingsMap.putAll(getTransportProfileSSLSettings(settings));
-        if (TcpTransport.isUntrustedRemoteClusterEnabled()) {
-            sslSettingsMap.put(XPackSettings.REMOTE_CLUSTER_SSL_PREFIX, settings.getByPrefix(XPackSettings.REMOTE_CLUSTER_SSL_PREFIX));
+        if (REMOTE_CLUSTER_PORT_ENABLED.get(settings)) {
+            sslSettingsMap.put(XPackSettings.REMOTE_CLUSTER_SSL_PREFIX, getRemoteClusterSslSettings(settings));
         }
```
Comment (ywangd) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/SSLService.java:597]: Changed this test to use `remote_cluster.enabled` setting value instead of the feature flag. If the feature falg is not enabled, the setting won't be registered and hence this check works the same as checking the feature flag itself in production code. However, checking the setting value makes it possible to test the behaviour in both snapshot and release tests.

**Comment by ywangd** [comment]
Failures are unrelated and I raised https://github.com/elastic/elasticsearch/issues/93335

File: x-pack/plugin/security/qa/multi-cluster/build.gradle [hunk 4]
```diff
@@ -6,76 +6,88 @@
...
-  setting 'xpack.security.remote_cluster.ssl.verification_mode', 'none'
-  setting 'xpack.security.remote_cluster.ssl.client_authentication', 'none'
+  // Server side SSL configuration for remote cluster
+  extraConfigFile 'remote-cluster.key', file("src/test/resources/ssl/remote_cluster.key")
+  extraConfigFile 'remote-cluster.cert', file("src/test/resources/ssl/remote_cluster.crt")
+  extraConfigFile 'remote-cluster.ca', file("src/test/resources/ssl/remote-cluster-ca.crt")
+
+  setting 'xpack.security.remote_cluster.ssl.enabled', 'true'
+  setting 'xpack.security.remote_cluster.ssl.key', 'remote-cluster.key'
+  setting 'xpack.security.remote_cluster.ssl.certificate', 'remote-cluster.cert'
```
Comment (jakelandis) [inline x-pack/plugin/security/qa/multi-cluster/build.gradle:46]: is the certificated needed in the fulfilling cluster ?

**Comment by ywangd** [comment]
@elasticmachine update branch

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/core/security/transport/ProfileConfigurations.java [hunk 5]
```diff
@@ -8,25 +8,61 @@
...
+import java.util.HashMap;
 import java.util.Map;
 import java.util.Set;
 
 import static org.elasticsearch.transport.RemoteClusterPortSettings.REMOTE_CLUSTER_PORT_ENABLED;
 import static org.elasticsearch.transport.RemoteClusterPortSettings.REMOTE_CLUSTER_PROFILE;
+import static org.elasticsearch.xpack.core.XPackSettings.REMOTE_CLUSTER_SSL_ENABLED;
 import static org.elasticsearch.xpack.core.security.SecurityField.setting;
 
 public final class ProfileConfigurations {
```
Comment (jakelandis) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/core/security/transport/ProfileConfigurations.java:24]: Can you add some java doc to this class ? Specifically calling out the difference between transport internal, transport for legacy remote cluster connections and now the new dedicated remote cluster port that is backed but not configurable via the named profile.

File: x-pack/plugin/security/qa/multi-cluster/build.gradle [hunk 6]
```diff
@@ -6,76 +6,88 @@
...
-configurations {
-  signedCerts
-  rootCert
-}
-
-dependencies {
-  signedCerts project(path: ':x-pack:plugin:core', configuration: 'signedCerts')
-  rootCert project(path: ':x-pack:plugin:core', configuration: 'rootCert')
-}
-
```
Comment (jakelandis) [inline x-pack/plugin/security/qa/multi-cluster/build.gradle:23]: any reason why the existing certs are not sufficient ? Just trying to avoid more cert explosion in the source code.

File: x-pack/plugin/security/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/remotecluster/RemoteCluster2SecuritySmokeIT.java [hunk 7]
```diff
@@ -18,7 +21,7 @@
  * This test suite will be run twice: Once against the fulfilling cluster, then again against the querying cluster. The typical usage is to
  * conditionalize on whether the test is running against the fulfilling or the querying cluster.
  */
-public class RemoteClusterSecuritySmokeIT extends ESRestTestCase {
+public class RemoteCluster2SecuritySmokeIT extends ESRestTestCase {
```
Comment (jakelandis) [inline x-pack/plugin/security/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/remotecluster/RemoteCluster2SecuritySmokeIT.java:24]: nit: we should probably be consistent with naming and I would prefer to avoid using "2" to mean the newer model.  I prefer (internally) referring to 1.0 model as legacy and no qualifiers for the new model.

**Review by jakelandis — COMMENTED** [review]
changes are looking good and I should be able to finish the review tomorrow but in the meantime a couple minor questions.

File: x-pack/plugin/security/qa/multi-cluster/build.gradle (hunk 4 above)
Comment (ywangd) [reply x-pack/plugin/security/qa/multi-cluster/build.gradle:46]: Yes the server needs a signed certficiate which cannot be derived from just the private key. Our code [prevents node from starting up](https://github.com/elastic/elasticsearch/blob/b4484cd8c62b7080c8038d292975e03fbc03a202/libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/SslConfigurationLoader.java#LL352C4-L352C4) if `key` is configured without a `certificate`.

NOTE: the linked code actually has a bug in the error message: It should report key_path instead of keystore_path. I'll have a separate PR to fix it.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/core/security/transport/ProfileConfigurations.java (hunk 5 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/core/security/transport/ProfileConfigurations.java:58]: That's a good suggestion. I added [javadoc](https://github.com/elastic/elasticsearch/pull/93334/commits/cd4cca93464202b34d5428cde2ee312175b22eb1#diff-4464d67e0e26bda7adba76f549b07d1e8cc1ddd122fa37b4edd40558b351884f) for both the class and the method. Please let me know whether it reads correct for you. Thanks!

File: x-pack/plugin/security/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/remotecluster/RemoteCluster2SecuritySmokeIT.java (hunk 7 above)
Comment (ywangd) [reply x-pack/plugin/security/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/remotecluster/RemoteCluster2SecuritySmokeIT.java]: I renamed the other `RemoteClusterSecuritySmokeIT` to `LegacyRemoteClusterSecuritySmokeIT` and removed `2` from this class. I wanted to differentiate between the two classes by names because IntelliJ sometimes gets confused between them and causes issues in debugging. Thanks!

File: x-pack/plugin/security/qa/multi-cluster/build.gradle (hunk 6 above)
Comment (ywangd) [reply x-pack/plugin/security/qa/multi-cluster/build.gradle:23]: The `signedCerts`, even when they are for different clusters, are all signed by the same CA. I think this is by design for testing the trust restriction. But it does not really work here because I'd like to have completely separate trust chains for transport and remote cluster interface (and perhaps http in future). The separate trusts are necessary to test that the explicit configurations are working as intended. The `testnode` cert/key from the `core` project is often used by other test suite for configuring the transport. I could use it here as well to reduce a set of key/cert/ca. But I'd still need one set for the remote cluster. I didn't want to use the `httpCa.p12` from the core because the p12 format means the test cannot run in FIPS. It is also not really fit for the purpose since it is meant to be http. Overall, I think reusing existing certs is a bit problematic and could be confusing at times. Hence I went ahead to create dedicated ones. It's not a great solution. But it should provide better clarity till we have a better, systematic way to reuse key/cert/ca (or generate them on the fly). I hope this makes sense.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/core/security/transport/ProfileConfigurations.java [hunk 8]
```diff
@@ -8,25 +8,106 @@
...
+     *
+     * @param settings Settings of the ES node
+     * @param sslService For resolving the SSL configuration based on its prefix
+     * @param sslEnabledOnly If true, only include the SSL configuration if SSL is enabled for the profile.
+     *                       If false, SSL configuration is included for a profile regardless whether SSL is actually enabled for it.
+     * @return A map that contains {@link SslConfiguration} for each named transport profile as well
+     *         as an entry for the "default" profile. If the remote_cluster feature is enabled, it also
+     *         contains an entry for the synthetic "_remote_cluster" profile.
+     */
+    public static Map<String, SslConfiguration> get(Settings settings, SSLService sslService, boolean sslEnabledOnly) {
```
Comment (jakelandis) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/core/security/transport/ProfileConfigurations.java:73]: I find `sslEnabledOnly`  hard to follow due to the intersection with the 2 known profiles. My expectations due to the naming would be to return all profiles with ssl enabled irregardless if a known name or not. As is (assuming i am reading this correctly) - if I pass in true and the remote cluster port ssl is enabled but the transport ssl is not then I only get remote cluster profile ... but if I pass in false for the sslEnabledOnly flag then I still get back the remote cluster profile but also will get back custom and default profiles irregadless if they are enabled or not.  I think this method is trying to be too smart and is complicated by the boolean logic of three independent booleans that conditionally change the output.  I think I am partially confused since I don't understand why we need those conditionals at all .

Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/core/security/transport/ProfileConfigurations.java:73]: The existing method return profiles regardless whether SSL is enabled for them. It also returns _remote_cluster synthetic profile as long as the feature is enabled (whether SSL is enabled is irrelevant). The similar effect to the `sslEnabledOnly` parameter is done by the caller, i.e.:

For `SecurityNett4Transport`, it is
```java
if (transportSslEnabled) {
    profiles = ProfileConfigurations.get(...);
} else {
    profiles = Map.of();
}
```

For `SecurityServerTransportInterceptor`, it is just
```
profiles = ProfileConfigurations.get(...);
```
because it needs initialize ServerTransportFilter for each filter regardless whether SSL is enabled for it.

With the addition of `remoteClusterSslEnabled`, if we keep the existing pattern, for `SecurityNett4Transport`, the code would be something like
```java
profiles = new HashMap<>();
if (transportSslEnabled) {
    profiles.putAll(ProfileConfigurations.getProfilesForDefaultAndTransportProfiles(...));
} 
if (remoteClusterSslEnabled) {
    assert profiles.containsKey(`_remote_cluster`) == false;
    profiles.putAll(ProfileConfigurations.getProfilesForRemoteCluster(...));
}
profiles = Map.copyOf(profiles);
```

So my choice was between (1) split the `ProfileConfigurations.get` method into 2 or (2) keep the single method but move the condition check inside it. I decided to go with the later because it seems to be less change from the API perspective. It also keep the consistency check (assertions) inside the method. But I am also OK to go with the former approach. Or alternatively, we can let `ProfileConfigurations.get` always builds all profiles and remove entries from the Map in the callers. 

What do you think? Please let me know. Thanks!

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/SSLService.java [hunk 9]
```diff
@@ -656,6 +647,35 @@ private void validateServerConfiguration(String prefix) {
...
     }
 
+    private void maybeValidateRemoteClusterServerConfiguration() {
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/SSLService.java:650]: I can change it to `== false`. 
Some explanation on this style: I personally prefer this style because I feel it reads better. Noramlly a negate check is `!condition` and I read it as "not condition". Since we don't use `!` for negate, I find the style `false == condition` reads the same way for me. As soon as I see the leading `false`, I am reading it as "not condition". But if the `false` is at the end, I find it is less readable because I don't know it is a negate till the end.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/core/security/transport/ProfileConfigurations.java (hunk 8 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/core/security/transport/ProfileConfigurations.java:73]: Just to be clear, my preferred approaches, from most preferred to least, are:
1. Single method with the boolean parameter as is in the PR
2. Split the method into 2 as described above
3. Single method that always build all profiles and let callers remove unwanted entries 

Please let me know if any of these works for you. Or if you have other suggestions. Thanks!

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/SSLService.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/SSLService.java:650]: Pushed https://github.com/elastic/elasticsearch/pull/93334/commits/29185f39e8364e63d7e83a620397a58f1b3cf302

**Comment by ywangd** [comment]
@elasticmachine run elasticsearch-ci/release-tests

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptorTests.java [hunk 10]
```diff
@@ -833,6 +841,104 @@ public TransportResponse read(StreamInput in) {
...
+            new DestructiveOperations(
+                Settings.EMPTY,
+                new ClusterSettings(Settings.EMPTY, Collections.singleton(DestructiveOperations.REQUIRES_NAME_SETTING))
+            ),
+            new RemoteClusterAuthorizationResolver(settings, clusterService.getClusterSettings())
+        );
+
+        final Map<String, ServerTransportFilter> profileFilters = securityServerTransportInterceptor.getProfileFilters();
+        assertThat(profileFilters.keySet(), contains("default"));
+        assertThat(profileFilters.get("default").isExtractClientCert(), is(transportSslEnabled));
```
Comment (jakelandis) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptorTests.java:939]: nit: can you assert no _remote_cluster profileFilter (or containsOnly(default))

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/core/security/transport/ProfileConfigurationsTests.java [hunk 11]
```diff
@@ -50,25 +56,134 @@ public void testGetInsecureTransportProfileConfigurations() {
...
         final SslConfiguration defaultConfig = sslService.getSSLConfiguration("xpack.security.transport.ssl");
-        final Map<String, SslConfiguration> profileConfigurations = ProfileConfigurations.get(settings, sslService, defaultConfig);
+        final Map<String, SslConfiguration> profileConfigurations = ProfileConfigurations.get(settings, sslService, true);
         assertThat(profileConfigurations.size(), Matchers.equalTo(2));
         assertThat(profileConfigurations.keySet(), Matchers.containsInAnyOrder("none", "default"));
         assertThat(profileConfigurations.get("none").verificationMode(), Matchers.equalTo(SslVerificationMode.NONE));
         assertThat(profileConfigurations.get("default"), Matchers.sameInstance(defaultConfig));
     }
 
+    public void testTransportAndRemoteClusterSslCanBeEnabledIndependently() {
```
Comment (jakelandis) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/core/security/transport/ProfileConfigurationsTests.java:66]: nit: a custom profile with a comment here stating that custom profiles are only enabled/disabled with the default profile might help future devs understand some intent

**Review by jakelandis — APPROVED** [review]
LGTM (feel free to ignore any nits:) 
Thanks for the additional explanation in person.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptorTests.java (hunk 10 above)
Comment (ywangd) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptorTests.java:939]: The names of Hamcrest matcher methods can be confusing. In this case, the `contains` method asserts that the two collections are both (1) the same length and (2) elements are equal at every position. So it is already a containsOnly and more (because it also asserts order). The method for asserting a subset matching is `hasItems`.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/SSLService.java [hunk 12]
```diff
@@ -656,6 +647,35 @@ private void validateServerConfiguration(String prefix) {
...
+        if (REMOTE_CLUSTER_PORT_ENABLED.get(settings) == false) {
+            return;
+        }
+        final String prefix = "xpack.security.remote_cluster.ssl";
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/SSLService.java:654]: There is some inconsistency how we set the constants for these prefix. The constant `REMOTE_CLUSTER_SSL_PREFIX` has an ending dot. But the string here does not have it. Some code accepts both forms (with or without ending dot), but other code works well for only one of the forms. This inconsistency is carried (modelled) after the existing prefix (transport, http) and they also use separate strings in the relevant methods. 

In summary, I'll keep it as is for this PR and have a separate PR to address the inconsistency wholisticly and drop the unncessary string literals.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/core/security/transport/ProfileConfigurationsTests.java (hunk 11 above)
Comment (ywangd) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/core/security/transport/ProfileConfigurationsTests.java:66]: Good point. I added a comment to the main code `ProfileConfigurations` for this. I think it is more useful over there. See 8e92bff1c307010c3eec4de23301a2d8b1af94d3

**Comment by ywangd** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by ywangd** [comment]
The release-tests failed again with downloading error
> Could not get resource 'https://.../downloads/elasticsearch/elasticsearch-7.15.2-linux-x86_64.tar.gz'.

They take very long to run (over 2 hours) and failed 2 times with above error. I am going to ignore it and merge the PR since my other PR (#93409) depends on it. The release-tests here are just nice to have. Its own CI runs separately as a periodical job. So we are not losing any cover.
